src/main/java/com/unascribed/fabrication/mixin/i_woina/void_fog_particles/MixinClientWorld.java

```diff
@@ -0,0 +1,57 @@
+package com.unascribed.fabrication.mixin.i_woina.void_fog_particles;
+
+import com.unascribed.fabrication.FabConf;
+import com.unascribed.fabrication.support.EligibleIf;
+import com.unascribed.fabrication.support.Env;
+import com.unascribed.fabrication.util.DepthSuspendParticle;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.profiler.Profiler;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.world.MutableWorldProperties;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.function.Supplier;
+
+@Mixin(ClientWorld.class)
+@EligibleIf(configAvailable="*.void_fog_particles", envMatches=Env.CLIENT)
+public abstract class MixinClientWorld extends World {
+
+	@Shadow
+	@Final
+	private MinecraftClient client;
+
+	protected MixinClientWorld(MutableWorldProperties properties, RegistryKey<World> registryRef, DimensionType dimensionType, Supplier<Profiler> profiler, boolean isClient, boolean debugWorld, long seed) {
+		super(properties, registryRef, dimensionType, profiler, isClient, debugWorld, seed);
+	}
+
+	//Ported from 1.7.10
+	//note that the source did not render the particles in flat worlds, however this is client-side so that can't be checked
+	//also note that i'm trash at locating mappings so any of the fairly many assumptions made could be wrong
+	@Inject(at=@At("HEAD"), method="doRandomBlockDisplayTicks(III)V")
+	public void voidParticles(int centerX, int centerY, int centerZ, CallbackInfo ci){
+		if (!FabConf.isEnabled("*.void_fog_particles") || this.getDimension().hasCeiling()) return;
+		int floor = this.getDimension().getMinimumY();
+		if (floor+24<centerY) return;
+		for(int i=0; i<1000; ++i) {
+			int x = centerX + this.random.nextInt(16) - this.random.nextInt(16);
+			int y = centerY + this.random.nextInt(16) - this.random.nextInt(16);
+			int z = centerZ + this.random.nextInt(16) - this.random.nextInt(16);
+			//Source checked for Material.AIR, assuming this is the same
+			if(this.isAir(new BlockPos(x, y, z))) {
```

Comment:
This for loop runs way too many times for this to be okay. Construct a BlockPos.Mutable outside the loop and `set` it to the coordinates.